File: creditline-decision-agent-openai/scripts/run_demo.py
"""End-to-end demo for the OpenAI build: one escalated, human-overridden decision.

Drives the SAME `mimic-creditline` MCP tools (no LLM / API key required) through
the full flow for the seeded escalation case, recording the OpenAI agent's
provenance, then simulates the officer override. Proves the environment +
provenance wiring without an OpenAI key.

    python -m scripts.run_demo

The live, model-driven version is `python -m agent.main`.
"""
from __future__ import annotations

import asyncio
import os
import sys
from decimal import Decimal
from pathlib import Path

import asyncpg
from dotenv import load_dotenv

load_dotenv()
os.environ.setdefault("APPROVAL_POLL_SECONDS", "1")
os.environ.setdefault("APPROVAL_TIMEOUT_SECONDS", "60")

# Reuse the environment layer (MCP server + officer console) from the sibling.
# Append (don't insert) so THIS project's `agent` package wins; only the
# sibling-only `mcp_servers` package is resolved from there.
SIBLING = Path(__file__).resolve().parents[2] / "creditline-decision-agent"
sys.path.append(str(SIBLING))

import importlib.util  # noqa: E402

from agent.decision import evaluate  # noqa: E402  (this project's copy)
from agent.prompts import MODEL_VERSION, PROMPT_VERSION_HASH  # noqa: E402
from mcp_servers.mimic_creditline import server as S  # noqa: E402  (from sibling)


# Load the sibling's officer console by path (both projects have a `scripts`
# package, so a normal import would collide with this project's own).
def _load_sibling_officer():
    spec = importlib.util.spec_from_file_location(
        "sibling_officer_console", str(SIBLING / "scripts" / "officer_console.py")
    )
    mod = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(mod)
    return mod


_officer = _load_sibling_officer()
_admin_dsn = _officer._admin_dsn
resolve = _officer.resolve

DEMO_REQUEST_ID = "11111111-1111-1111-1111-111111111111"


async def _approval_id_for(decision_id: str) -> str:
    conn = await asyncpg.connect(_admin_dsn())
    try:
        for _ in range(30):
            row = await conn.fetchrow(
                "SELECT approval_id FROM approvals WHERE decision_id = $1 "
                "ORDER BY requested_at DESC LIMIT 1",
                __import__("uuid").UUID(decision_id),
            )
            if row:
                return str(row["approval_id"])
            await asyncio.sleep(0.5)
    finally:
        await conn.close()
    raise RuntimeError("approval row never appeared")


async def main() -> None:
    print("=== Credit-Line Decision Agent (OpenAI Agents SDK) — deterministic demo ===\n")

    req = await S.creditline_get_request(DEMO_REQUEST_ID)
    cust = await S.creditline_get_customer(req["customer_id"])
    bureau = await S.creditline_pull_bureau(req["customer_id"])
    policy = await S.creditline_get_active_policy()
    customer = cust["customer"]

    print(f"Applicant       : {customer['full_name']}")
    print(f"Request          : {req['request_type']} → limit {req['requested_limit']}")
    print(f"Active policy     : {policy['version']}  thresholds {policy['thresholds']}\n")

    rec = evaluate(
        requested_limit=Decimal(str(req["requested_limit"])),
        annual_income=Decimal(str(customer["annual_income"])),
        bureau=bureau,
        policy_thresholds=policy["thresholds"],
    )
    print(f"Computed DTI      : {rec.dti:.3f}")
    print(f"Recommendation    : {rec.outcome.upper()}")
    for r in rec.reasons:
        print(f"   • {r}")
    print()

    decision = await S.creditline_record_decision(
        request_id=DEMO_REQUEST_ID,
        outcome=rec.outcome,
        rationale="; ".join(rec.reasons),
        model_version=MODEL_VERSION,
        prompt_version_hash=PROMPT_VERSION_HASH,
        policy_version=policy["version"],
        bureau_report_id=bureau["bureau_report_id"],
        approved_limit=float(rec.approved_limit) if rec.approved_limit is not None else None,
    )
    print(f"Recorded decision : {decision['decision_id']}  "
          f"requires_human_approval={decision['requires_human_approval']}")
    print(f"   model_version       = {MODEL_VERSION}")
    print(f"   prompt_version_hash = {PROMPT_VERSION_HASH}\n")

    if not decision["requires_human_approval"]:
        print("Auto-approved — no human gate. Done.")
        return

    print("Opening human-in-the-loop approval gate (agent now blocks)…")
    summary = (
        f"{customer['full_name']} requests {req['requested_limit']}. "
        f"Above ceiling and DTI {rec.dti:.3f} over appetite. Recommend escalate."
    )
    gate = asyncio.create_task(
        S.creditline_request_approval(decision_id=decision["decision_id"], summary=summary)
    )

    async def officer():
        approval_id = await _approval_id_for(decision["decision_id"])
        await asyncio.sleep(1)
        print("  [officer] reviewing… modifying to an approved limit of 12,000")
        await resolve(
            approval_id=approval_id,
            action="modify",
            approver_id="co-114-jmalik",
            justification="Strong relationship, clean record; approve reduced 12,000 limit.",
            modified_limit=12000,
        )

    officer_task = asyncio.create_task(officer())
    result = await gate
    await officer_task

    print("\nApproval resolved:")
    print(f"   status        : {result['status']}")
    print(f"   is_override   : {result['is_override']}  ← Art.14 / SR 11-7 effective challenge")
    print(f"   approved limit: {result['modified_limit']}")
    print(f"   approver       : {result['approver_id']} ({result['approver_role']})")

    await S.creditline_notify_customer(
        request_id=DEMO_REQUEST_ID,
        outcome="approved",
        approved_limit=result["modified_limit"],
        idempotency_key=decision["decision_id"],
    )
    print("\n=== demo complete: escalated → human override → approved at 12,000 ===")


if __name__ == "__main__":
    asyncio.run(main())

File: creditline-decision-agent-openai-ts/src/scripts/run-demo.ts
// ============================================================================
// Deterministic end-to-end demo for the OpenAI Agents build: one escalated,
// human-overridden decision (no LLM / API key). Drives the SAME
// `mimic-creditline` MCP server (spawned from the sibling base project) over
// stdio through the full flow for the seeded escalation case, recording the
// OpenAI agent's provenance, then simulates the officer override.
//
//   npm run demo
//
// The live, model-driven version is `npm run agent`. The shared, interactive
// credit-officer console lives in the base project (`npm run officer` there);
// this demo inlines the same admin-side resolution for a hands-off run.
// ============================================================================
import { fileURLToPath } from "node:url";
import { dirname, join, resolve as pathResolve } from "node:path";
import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { StdioClientTransport } from "@modelcontextprotocol/sdk/client/stdio.js";
import pg from "pg";
import { evaluate } from "../agent/decision.js";
import { MODEL_VERSION, PROMPT_VERSION_HASH } from "../agent/prompts.js";

try { process.loadEnvFile(); } catch { /* env may come from the shell */ }

pg.types.setTypeParser(1700, (v) => (v === null ? null : Number(v))); // numeric → number

const __dirname = dirname(fileURLToPath(import.meta.url));
const PROJECT_ROOT = pathResolve(__dirname, "..", "..");
const SIBLING = pathResolve(PROJECT_ROOT, "..", "creditline-decision-agent-ts");
const SERVER_PATH = join(SIBLING, "src", "mcp", "server.ts");
const TSX_BIN = join(PROJECT_ROOT, "node_modules", ".bin", "tsx");

const DEMO_REQUEST_ID = "11111111-1111-1111-1111-111111111111";
const AGENT_DSN =
  process.env.MIMIC_DB_DSN || "postgresql://creditline_agent:agent_demo_pw@localhost:5433/mimic_creditline";

function adminDsn(): string {
  const host = process.env.ADMIN_DB_HOST || "localhost";
  const port = process.env.ADMIN_DB_PORT || "5433";
  const user = process.env.ADMIN_DB_USER || "postgres";
  const pw = process.env.ADMIN_DB_PASSWORD || "postgres";
  return `postgresql://${user}:${pw}@${host}:${port}/mimic_creditline`;
}

async function call(client: Client, name: string, args: Record<string, unknown>): Promise<any> {
  const res: any = await client.callTool({ name, arguments: args });
  return JSON.parse(res.content?.[0]?.text ?? "{}");
}

// The officer's admin-side override (modify to a reduced approved limit).
async function officerModify(decisionId: string, modifiedLimit: number): Promise<void> {
  const c = new pg.Client({ connectionString: adminDsn() });
  await c.connect();
  try {
    let approvalId: string | undefined;
    for (let i = 0; i < 30 && !approvalId; i++) {
      const r = await c.query(
        `SELECT approval_id FROM approvals WHERE decision_id = $1 ORDER BY requested_at DESC LIMIT 1`,
        [decisionId],
      );
      if (r.rowCount) approvalId = r.rows[0].approval_id;
      else await new Promise((res) => setTimeout(res, 500));
    }
    if (!approvalId) throw new Error("approval row never appeared");
    await new Promise((res) => setTimeout(res, 1000));
    console.log("  [officer] reviewing… modifying to an approved limit of 12,000");
    const appr = (
      await c.query(`SELECT d.decision_id, d.request_id FROM approvals a
                       JOIN decisions d ON d.decision_id = a.decision_id
                      WHERE a.approval_id = $1`, [approvalId])
    ).rows[0];
    await c.query("BEGIN");
    await c.query(
      `UPDATE approvals SET status='modified', is_override=true, modified_limit=$2,
              approver_id='co-114-jmalik', justification=$3, resolved_at=now()
        WHERE approval_id=$1`,
      [approvalId, modifiedLimit, "Strong relationship, clean record; approve reduced 12,000 limit."],
    );
    await c.query(`UPDATE decisions SET outcome='approved', approved_limit=$2 WHERE decision_id=$1`, [
      appr.decision_id,
      modifiedLimit,
    ]);
    await c.query(`UPDATE credit_line_requests SET status='approved' WHERE request_id=$1`, [appr.request_id]);
    await c.query("COMMIT");
  } catch (err) {
    await c.query("ROLLBACK").catch(() => {});
    throw err;
  } finally {
    await c.end();
  }
}

async function main() {
  console.log("=== Credit-Line Decision Agent (OpenAI Agents SDK) — deterministic demo ===\n");

  const transport = new StdioClientTransport({
    command: TSX_BIN,
    args: [SERVER_PATH],
    cwd: SIBLING,
    env: {
      ...process.env,
      MIMIC_DB_DSN: AGENT_DSN,
      APPROVAL_POLL_SECONDS: "1",
      APPROVAL_TIMEOUT_SECONDS: "60",
    } as Record<string, string>,
  });
  const client = new Client({ name: "demo", version: "0.1.0" });
  await client.connect(transport);

  const req = await call(client, "creditline_get_request", { request_id: DEMO_REQUEST_ID });
  const cust = await call(client, "creditline_get_customer", { customer_id: req.customer_id });
  const bureau = await call(client, "creditline_pull_bureau", { customer_id: req.customer_id });
  const policy = await call(client, "creditline_get_active_policy", {});
  const customer = cust.customer;

  console.log(`Applicant        : ${customer.full_name}`);
  console.log(`Request          : ${req.request_type} → limit ${req.requested_limit}`);
  console.log(`Active policy    : ${policy.version}  thresholds ${JSON.stringify(policy.thresholds)}\n`);

  const rec = evaluate({
    requested_limit: Number(req.requested_limit),
    annual_income: Number(customer.annual_income),
    bureau,
    policy_thresholds: policy.thresholds,
  });
  console.log(`Computed DTI     : ${rec.dti.toFixed(3)}`);
  console.log(`Recommendation   : ${rec.outcome.toUpperCase()}`);
  for (const r of rec.reasons) console.log(`   • ${r}`);
  console.log();

  const decision = await call(client, "creditline_record_decision", {
    request_id: DEMO_REQUEST_ID,
    outcome: rec.outcome,
    rationale: rec.reasons.join("; "),
    model_version: MODEL_VERSION,
    prompt_version_hash: PROMPT_VERSION_HASH,
    policy_version: policy.version,
    bureau_report_id: bureau.bureau_report_id,
    approved_limit: rec.approved_limit,
  });
  console.log(`Recorded decision: ${decision.decision_id}  requires_human_approval=${decision.requires_human_approval}`);
  console.log(`   model_version       = ${MODEL_VERSION}`);
  console.log(`   prompt_version_hash = ${PROMPT_VERSION_HASH}\n`);

  if (!decision.requires_human_approval) {
    console.log("Auto-approved — no human gate. Done.");
    await client.close();
    return;
  }

  console.log("Opening human-in-the-loop approval gate (agent now blocks)…");
  const summary =
    `${customer.full_name} requests ${req.requested_limit}. Above ceiling and ` +
    `DTI ${rec.dti.toFixed(3)} over appetite. Recommend escalate.`;

  const gate = call(client, "creditline_request_approval", { decision_id: decision.decision_id, summary });
  const officer = officerModify(decision.decision_id, 12000);
  const [gateResult] = await Promise.all([gate, officer]);

  console.log("\nApproval resolved:");
  console.log(`   status        : ${gateResult.status}`);
  console.log(`   is_override   : ${gateResult.is_override}  ← Art.14 / SR 11-7 effective challenge`);
  console.log(`   approved limit: ${gateResult.modified_limit}`);
  console.log(`   approver      : ${gateResult.approver_id} (${gateResult.approver_role})`);

  await call(client, "creditline_notify_customer", {
    request_id: DEMO_REQUEST_ID,
    outcome: "approved",
    approved_limit: gateResult.modified_limit,
    idempotency_key: decision.decision_id,
  });

  await client.close();
  console.log("\n=== demo complete: escalated → human override → approved at 12,000 ===");
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
